src/libcollections/str.rs

```diff
@@ -1634,30 +1634,53 @@ impl str {
     #[stable(feature = "unicode_case_mapping", since = "1.2.0")]
     pub fn to_lowercase(&self) -> String {
         let mut s = String::with_capacity(self.len());
-        for (i, c) in self[..].char_indices() {
-            if c == 'Σ' {
-                // Σ maps to σ, except at the end of a word where it maps to ς.
-                // This is the only conditional (contextual) but language-independent mapping
-                // in `SpecialCasing.txt`,
-                // so hard-code it rather than have a generic "condition" mechanism.
-                // See https://github.com/rust-lang/rust/issues/26035
-                map_uppercase_sigma(self, i, &mut s)
-            } else {
-                s.extend(c.to_lowercase());
+        let mut left = None;
+
+        // Try to collect slices of lower case characters to push into the
+        // result or extend with the lower case version if an upper case
+        // character is found.
+        for (i, ch) in self.char_indices() {
+            if !ch.is_lowercase() && ch.is_alphabetic() {
```

Comment:
`ch.to_lowercase` would return an `Iterator`, so I'd have to create it, make it peekable, check the first character is different and then use it.

The issue is it would defeat the purpose of the optimization, because the conversion is what takes most of the time.

If someone with Unicode knowledge can chime in it would be nice, I'm wary about it too.

From what I understand non-alphabetic characters can't be turned to lower case, and the logic is that some characters (title case and non-alphabetic ones) are neither upper case nor lower case.